### PR TITLE
PATRON2020-325 gateway request validation

### PR DESCRIPTION
**Pull request by ArturKos**

### CODE REVIEW

File: gateway/dashboard.h [hunk 1]
```diff
@@ -127,35 +146,50 @@ class Dashboard
...
-                hvac_rooms[i].heating_temperature = request["heatingTemperature"].get<int>();
-                hvac_rooms[i].cooling_temperature = request["coolingTemperature"].get<int>();
+                if(hvac_rooms[i].temperature_sensor_id==0)
+                    if(!(request.contains("temperatureSensorId")))
+                        return 1;
+                    else
+                        hvac_rooms[i].temperature_sensor_id = request["temperatureSensorId"].get<int>();      
+
+                hvac_rooms[i].heating_temperature = request["heating-temperature"].get<int>();
+                hvac_rooms[i].cooling_temperature = request["cooling-temperature"].get<int>();
```
Comment (kuba8686) [inline gateway/dashboard.h]: proper JSON fields for heating and cooling temperature is: heatingTemperature oraz coolingTemperature. Thats the filds that is provided at sensor.h (void to_json, void to_json - have to be corrected). That fields names are already implemented by JS and Android. You also have to use it like that.

File: gateway/request_validation.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      return true;
+  return false;
+}
+
+bool validate_hvac_request(json hvac_json){
+    if(hvac_json.empty())
+        return true;
+    if(!(hvac_json.contains("heating-temperature")))
+        return true;
+    if(!(hvac_json.contains("cooling-temperature")))
```
Comment (kuba8686) [inline gateway/request_validation.h:50]: as above

File: gateway/request_validation.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        return true;
+    if(hvac_json["cooling-temperature"].get<int>()>400)
+        return true;
+    if(hvac_json["hysteresis"].get<int>()<5)
+        return true;
+    if(hvac_json["hysteresis"].get<int>()>20)
+        return true;
+    if(hvac_json["heating-temperature"].get<int>()>hvac_json["cooling-temperature"].get<int>()-hvac_json["hysteresis"].get<int>())
+        return true;
+    if(hvac_json["cooling-temperature"].get<int>()<hvac_json["heating-temperature"].get<int>()+hvac_json["hysteresis"].get<int>())
```
Comment (kuba8686) [inline gateway/request_validation.h:68]: as above.

File: gateway/gateway.cpp [hunk 4]
```diff
@@ -1,144 +1,234 @@
+//#define CROW_ENABLE_SSL
```
Comment (kuba8686) [inline gateway/gateway.cpp:1]: its good for localhost testing, but for develop and server - ssl should be turned on.

File: gateway/dashboard.h (hunk 1 above)
Comment (ArturKos) [reply gateway/dashboard.h]: Please view https://intive.github.io/patronage20-concept/#get-/dashboard

Comment (ArturKos) [reply gateway/dashboard.h]: done

File: gateway/request_validation.h (hunk 2 above)
Comment (ArturKos) [reply gateway/request_validation.h]: done

File: gateway/request_validation.h (hunk 3 above)
Comment (ArturKos) [reply gateway/request_validation.h]: done

File: gateway/gateway.cpp (hunk 4 above)
Comment (ArturKos) [reply gateway/gateway.cpp]: done

File: gateway/dashboard.h (hunk 1 above)
Comment (kuba8686) [reply gateway/dashboard.h]: ok, You right. Your names is propper. It was innitial mistake at sensor.h - two different names, and the wrong one finally appear in dashboard. Its wrong due to schema but other groups implemented like that so we have to keep it :) Rest differences we will solve during working on /hvac request.

File: gateway/gateway.cpp [hunk 5]
```diff
@@ -152,26 +153,30 @@ int main(void)
...
-        pthread_rwlock_unlock(&q_rwlock);    
+        pthread_rwlock_unlock(&q_rwlock);
         return crow::response(200);
     });
 
     CROW_ROUTE(app, "/hvac").methods(crow::HTTPMethod::PUT)([&](const crow::request &req) {
         json hvac = json::parse(req.body);
-        if (hvac.empty())
+
+        if (validate_hvac_request(hvac))
```
Comment (cichy1024) [inline gateway/gateway.cpp:177]: It should be `if (!validate...)`. It looks like it returns error if validation succeeds.

File: gateway/gateway.cpp [hunk 6]
```diff
@@ -183,25 +188,26 @@ int main(void)
     });
 
     CROW_ROUTE(app, "/lights").methods(crow::HTTPMethod::PUT)([&](const crow::request &req) {
-        
+
         json light = json::parse(req.body);
-        
-        if (light.empty())
+
+        if (validate_light_request(light))
```
Comment (cichy1024) [inline gateway/gateway.cpp:194]: The same. Positive validation should mean it's OK.

File: gateway/request_validation.h [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+bool validate_light_request(json light_json){
+  if(light_json.empty())
+      return true;
+  if(!(light_json.contains("hue")))
+      return true;
+  if(!(light_json.contains("saturation")))
+      return true;
+  if(!(light_json.contains("value")))
+      return true;
```
Comment (cichy1024) [inline gateway/request_validation.h:17]: The assumption was we can change only one value.

File: gateway/request_validation.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+bool validate_hvac_request(json hvac_json){
+    if(hvac_json.empty())
+        return true;
+    if(!(hvac_json.contains("heatingTemperature")))
+        return true;
+    if(!(hvac_json.contains("coolingTemperature")))
+        return true;
+    if(!(hvac_json.contains("hysteresis")))
+        return true;
```
Comment (cichy1024) [inline gateway/request_validation.h:53]: The assumption was we can change only one value.

File: gateway/request_validation.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    if(hvac_json["coolingTemperature"].get<int>()>400)
+        return true;
+    if(hvac_json["hysteresis"].get<int>()<5)
+        return true;
+    if(hvac_json["hysteresis"].get<int>()>20)
+        return true;
+    if(hvac_json["heatingTemperature"].get<int>()>hvac_json["coolingTemperature"].get<int>()-hvac_json["hysteresis"].get<int>())
+        return true;
+    if(hvac_json["coolingTemperature"].get<int>()<hvac_json["heatingTemperature"].get<int>()+hvac_json["hysteresis"].get<int>())
+      return true;
```
Comment (cichy1024) [inline gateway/request_validation.h:69]: If we change only one value, dependencies should be verified against the dashboard content.

File: gateway/gateway.cpp (hunk 5 above)
Comment (ArturKos) [reply gateway/gateway.cpp]: done

File: gateway/gateway.cpp (hunk 6 above)
Comment (ArturKos) [reply gateway/gateway.cpp]: done

File: gateway/request_validation.h (hunk 7 above)
Comment (ArturKos) [reply gateway/request_validation.h]: done

File: gateway/request_validation.h (hunk 8 above)
Comment (ArturKos) [reply gateway/request_validation.h]: done

File: gateway/request_validation.h (hunk 9 above)
Comment (ArturKos) [reply gateway/request_validation.h]: done

File: gateway/dashboard.h [hunk 10]
```diff
@@ -127,35 +146,75 @@ class Dashboard
...
+                             return 1;
+                         hvac_rooms[i].cooling_temperature = request["coolingTemperature"].get<int>();
+                    }
+                    else if(request.contains("hysteresis")){
+                        if(hvac_rooms[i].heating_temperature>hvac_rooms[i].cooling_temperature-request["hysteresis"].get<int>())
+                            return 1;
+                        if(hvac_rooms[i].cooling_temperature<hvac_rooms[i].heating_temperature+request["hysteresis"].get<int>())
+                            return 1;
+                        hvac_rooms[i].hysteresis = request["hysteresis"].get<int>();
+                    }
```
Comment (cichy1024) [inline gateway/dashboard.h:197]: It's too complicated. You should use temporary values, assign to them values either stored or sent and then verify ones. What happens if we send heating temperature and cooling temperature but not hytheresis?

Comment (ArturKos) [reply gateway/dashboard.h]: done

File: gateway/dashboard.h [hunk 11]
```diff
@@ -127,35 +146,73 @@ class Dashboard
...
+                        hysteresis =  hvac_rooms[i].hysteresis;
+                    }
+
+                    if(tmp_heating>tmp_cooling-hysteresis)
+                        return 1;
+                    if(tmp_cooling<tmp_heating+hysteresis)
+                        return 1;
+                    hvac_rooms[i].heating_temperature = tmp_heating;
+                    hvac_rooms[i].cooling_temperature = tmp_cooling;
+                    hvac_rooms[i].hysteresis = hysteresis;
```
Comment (cichy1024) [inline gateway/dashboard.h:195]: Indentation!!! For a moment I thought it's a one big "if".

File: gateway/dashboard.h [hunk 12]
```diff
@@ -127,35 +146,73 @@ class Dashboard
...
-                hvac_rooms[i].heating_temperature = request["heatingTemperature"].get<int>();
-                hvac_rooms[i].cooling_temperature = request["coolingTemperature"].get<int>();
-                hvac_rooms[i].hysteresis = request["hysteresis"].get<int>();
-                hvac_rooms[i].temperature_sensor_id = request["temperatureSensorId"].get<int>();
-                hvac_rooms[i].window_sensor_ids = request["windowSensorIds"].get<std::vector<int>>();
+                if(hvac_rooms[i].temperature_sensor_id==0)
+                    if(!(request.contains("temperatureSensorId")))
+                        return 1;
+                    else
+                        hvac_rooms[i].temperature_sensor_id = request["temperatureSensorId"].get<int>();
```
Comment (cichy1024) [inline gateway/dashboard.h:166]: Please add curly brackets, so it would be obvious to which "if" the "else" relates.

**Pull request merged**